File: core/src/main/kotlin/org/pastalab/fray/core/concurrency/locks/CountDownLatchContext.kt
package org.pastalab.fray.core.concurrency.locks

import org.pastalab.fray.core.ThreadContext
import org.pastalab.fray.core.ThreadState
import org.pastalab.fray.core.concurrency.operations.CountDownLatchAwaitBlocking
import org.pastalab.fray.core.concurrency.operations.ThreadResumeOperation

class CountDownLatchContext(var count: Long) {
  val latchWaiters = mutableMapOf<Long, LockWaiter>()

  fun await(canInterrupt: Boolean, thread: ThreadContext): Boolean {
    if (count > 0) {
      if (canInterrupt) {
        thread.checkInterrupt()
      }
      latchWaiters[Thread.currentThread().id] = LockWaiter(canInterrupt, thread)
      return true
    }
    assert(count == 0L)
    return false
  }

  fun unblockThread(tid: Long, isTimeout: Boolean, isInterrupt: Boolean): Boolean {
    val lockWaiter = latchWaiters[tid] ?: return false
    if (isInterrupt && !lockWaiter.canInterrupt) {
      return false
    }
    val pendingOperation = lockWaiter.thread.pendingOperation
    assert(pendingOperation is CountDownLatchAwaitBlocking)
    lockWaiter.thread.pendingOperation = ThreadResumeOperation(!isTimeout)
    lockWaiter.thread.state = ThreadState.Enabled
    latchWaiters.remove(tid)
    if ((pendingOperation as CountDownLatchAwaitBlocking).timed) {
      return false
    } else {
      return true
    }
  }

  fun release(): Int {
    if (count == 0L) {
      return 0
    }
    count = 0
    var threads = 0
    for (lockWaiter in latchWaiters.values.toList()) {
      if (unblockThread(lockWaiter.thread.thread.id, false, false)) {
        threads += 1
      }
    }
    return threads
  }

  /*
   * Returns number of unblocked threads.
   */
  fun countDown(): Int {
    // If count is already zero we do not need to
    // do anything
    if (count == 0L) {
      return 0
    }
    count -= 1
    if (count == 0L) {
      var threads = 0
      for (lockWaiter in latchWaiters.values.toList()) {
        if (unblockThread(lockWaiter.thread.thread.id, false, false)) {
          threads += 1
        }
      }
      return threads
    }
    return 0
  }
}

File: dotnet/src/Fray/Core/Contexts/CountDownLatchContext.cs
using Fray.Core.Operations;

namespace Fray.Core.Contexts;

/// <summary>
/// Models a countdown latch (<see cref="Fray.FrayCountdownEvent"/>).
///
/// Mirrors <c>org.pastalab.fray.core.concurrency.context.CountDownLatchContext</c>.
/// </summary>
public sealed class CountDownLatchContext : IInterruptibleContext
{
    private readonly Dictionary<int, LockWaiter> _latchWaiters = new();

    public int LatchId { get; }
    public long Count { get; private set; }

    public CountDownLatchContext(long count, object latch)
    {
        LatchId = ObjectIds.Of(latch);
        Count = count;
    }

    /// <summary>Returns true when the calling thread has to block.</summary>
    public bool Await(bool canInterrupt, ThreadContext thread)
    {
        if (Count > 0)
        {
            if (canInterrupt)
            {
                thread.CheckInterrupt();
            }
            _latchWaiters[thread.Index] = new LockWaiter(canInterrupt, thread);
            return true;
        }
        return false;
    }

    /// <summary>Returns the number of threads unblocked by this countdown.</summary>
    public int CountDown()
    {
        if (Count == 0)
        {
            return 0;
        }
        Count -= 1;
        if (Count != 0)
        {
            return 0;
        }
        var unblocked = 0;
        foreach (var waiter in _latchWaiters.Values.ToList())
        {
            if (UnblockThread(waiter.Thread, InterruptionType.ResourceAvailable))
            {
                unblocked += 1;
            }
        }
        return unblocked;
    }

    public bool UnblockThread(ThreadContext thread, InterruptionType type)
    {
        if (type == InterruptionType.Force)
        {
            // Forced wakeup (deadlock recovery): open the latch entirely.
            while (Count > 0)
            {
                CountDown();
            }
            return false;
        }
        if (!_latchWaiters.TryGetValue(thread.Index, out var waiter))
        {
            return false;
        }
        if (type == InterruptionType.Interrupt && !waiter.CanInterrupt)
        {
            return false;
        }
        waiter.Thread.PendingOperation = new ThreadResumeOperation(type != InterruptionType.Timeout);
        waiter.Thread.State = FrayThreadState.Runnable;
        _latchWaiters.Remove(thread.Index);
        return true;
    }
}
